src/traces/sunburst/calc.js

```diff
@@ -0,0 +1,237 @@
+        //  which one is it?
+        getId = function(i) { return String(labels[i]); };
+    }
+
+    if(hasVals) len = Math.min(len, vals.length);
+
+    for(var i = 0; i < len; i++) {
+        if(isValid(i)) {
+            var id = getId(i);
+            var pid = parents[i] ? String(parents[i]) : '';
+
+            var cdi = {
+                i: i,
+                id: id,
+                pid: pid,
+                label: labels[i] ? String(labels[i]) : ''
+            };
+
+            if(hasVals) cdi.v = +vals[i];
+            cd.push(cdi);
+            addToLookup(pid, id);
+        }
+    }
+
+    if(!parent2children['']) {
+        var impliedRoots = [];
+        var k;
+        for(k in parent2children) {
+            if(!refs[k]) {
+                impliedRoots.push(k);
+            }
+        }
+
+        // if an `id` has no ref in the `parents` array,
+        // take it as being the root node
+
+        if(impliedRoots.length === 1) {
+            k = impliedRoots[0];
+            cd.unshift({
+                id: k,
+                pid: '',
+                label: k
+            });
+        } else {
+            return Lib.warn('Multiple implied roots, cannot build sunburst hierarchy.');
+        }
+    } else if(parent2children[''].length > 1) {
+        var dummyId = Lib.randstr();
+
+        // if multiple rows linked to the root node,
+        // add dummy "root of roots" node to make d3 build the hierarchy successfully
+
+        for(var j = 0; j < cd.length; j++) {
+            if(cd[j].pid === '') {
+                cd[j].pid = dummyId;
+            }
+        }
+
+        cd.unshift({
+            hasMultipleRoots: true,
+            id: dummyId,
+            pid: ''
+        });
+    }
+
+    // TODO might be better to replace stratify() with our own algorithm
+    var root;
+    try {
+        root = d3Hierarchy.stratify()
+            .id(function(d) { return d.id; })
+            .parentId(function(d) { return d.pid; })(cd);
+    } catch(e) {
+        return Lib.warn('Failed to build sunburst hierarchy.');
+    }
+
+    var hierarchy = d3Hierarchy.hierarchy(root);
+
+    if(hasVals) {
+        switch(trace.branchvalues) {
+            case 'extra':
+                hierarchy.sum(function(d) { return d.data.v; });
+                break;
+            case 'total':
+                hierarchy.each(function(d) {
+                    var v = d.data.data.v;
+
+                    if(d.children) {
+                        var partialSum = d.children.reduce(function(a, c) {
+                            return a + c.data.data.v;
+                        }, 0);
+                        if(v < partialSum) {
+                            d.value = partialSum;
+                            return Lib.warn([
+                                'Total value for node', d.data.data.id,
+                                'is smaller than the sum of its children.'
```

Comment:
There are two ways of denoting the size of sectors depending on `branchvalues`:
1- `total`: which is in absolute values
2- `extra`: which is relative to the sum of its children

Because we filter negative values, we can only reach this `Lib.warn` in `total`. In `extra`, we would have failed earlier when calling `stratify()`. At the moment, we handle the situation of parents that are smaller than children in two different manners.

I think the current behvaior of `extra` is more sane: it simply refuses to render something it can't correctly do. Until we add support for rendering parents that are smaller than the sum of its children (by padding the remaining space for example), I think it would make sense to bail out here to be correct _and_ consistent with `extra`.

Sorry for the long comment but I wanted to add context for other potential reviewers.
@plotly/plotly_js @nicolaskruchten

Reply:
> In `extra`, we would have failed earlier when calling `stratify()`
> I think the current behvaior of `extra` is more sane: it simply refuses to render something it can't correctly do

Hmm. I'm a little confused @antoinerg . Can you share an example of `stratify()` failing under `branchvalues:'extra'`?

Reply:
If you use a negative value in `extra`, we will filter that node out. If that node was a parent, `stratify()` will fail because it will be missing a node.

Reply:
Thanks very much for the tip @antoinerg !

Done in https://github.com/plotly/plotly.js/pull/3594/commits/4a2f7a5a59e7868b1a21222647d2632cb7d3edb0